File: python/turingTyping.py
import typing
import collections.abc

TuringSymbol = typing.TypeVar("TuringSymbol", bound=str)
TuringState = typing.TypeVar("TuringState", bound=str)

def raiseIfInvalidSymbol(symbol:TuringSymbol, name:str = "Symbol"):
    if len(symbol) != 1:
        raise ValueError(f"{name} must be a single character string but '{symbol}' has length {len(symbol)}")

def raiseIfUnknwodSymbol(symbol:TuringSymbol, symbols:collections.abc.Iterable[TuringSymbol],name:str = "Symbol"):
    if symbol not in symbols:
        raise ValueError(f"{name} '{symbol}' is not in the collection of valid symbols")

def raiseIfInvalidState(state:TuringState, name:str = "State"):
    if len(state) < 1:
        raise ValueError(f"{name} cannot be an empty string")
    
def raiseIfUnknwodState(state:TuringState, states:collections.abc.Iterable[TuringState],name:str = "State"):
    if state not in states:
        raise ValueError(f"{name} '{state}' is not in the collection of valid states")

File: cpp/turing-typing.cpp
#include <string>
#include <set>
#include <stdexcept>
#include <sstream>

#include "turing-typing.hpp"

void turing::throwIfStateIsInvalid(turing::turing_state &state, const char *name)
{
    if (state.size() == 0)
    {
        std::stringstream stream;
        stream << name << " cannot be an empty string.";
        throw std::invalid_argument(stream.str());
    }
}

void turing::throwIfStateIsUnknown(turing::turing_state &state, std::set<turing::turing_state> &states, const char *name)
{
    if (states.find(state) == states.end())
    {
        std::stringstream stream;
        stream << name << " '" << state << "' does not belong to the list of states.";
        throw std::invalid_argument(stream.str());
    }
}

void turing::throwIfSymbolIsUnknown(turing::turing_symbol &symbol, std::set<turing::turing_symbol> &symbols, const char *name)
{
    if (symbols.find(symbol) == symbols.end())
    {
        std::stringstream stream;
        stream << name << " '" << symbol << "' does not belong to the list of states.";
        throw std::invalid_argument(stream.str());
    }
}